src/storage/rect.rs

```diff
@@ -0,0 +1,632 @@
+use std::fmt::Debug;
+
+use glam::{IVec2, Vec2Swizzles};
+
+use crate::{Hex, HexLayout, OffsetHexMode, storage::HexStore};
+
+/// [`Vec`] Based storage for rectangular maps.
+///
+/// > See [this article](https://www.redblobgames.com/grids/hexagons/#map-storage)
+///
+/// [`RectMap`] is made for _rectangular_ large _dense_ maps, utilizing some
+/// tricks to map [`Hex`] coordinate to a positive 1D array.
+///
+/// It can be used only if:
+/// - The map is a rectanglar shape
+/// - The map is _dense_
+/// - No coordinate will be added or removed from the map
+///
+/// If your use case doesn't match all of the above, use a
+/// [`std::collections::HashMap`] instead
+///
+/// # Example
+/// ```rust
+/// use hexx::{
+///     storage::{HexStore, RectMap, RectMetadata, WrapStrategy},
+///     *,
+/// };
+///
+/// let layout = HexLayout::pointy()
+///     .with_hex_size(30.0)
+///     .with_origin(Vec2::ZERO);
+/// let rect_map = RectMetadata::default()
+///     .with_hex_layout(layout)
+///     .with_half_size(IVec2 { x: 8, y: 4 })
+///     .with_wrap_strategies([WrapStrategy::Cycle, WrapStrategy::Clamp])
+///     .build_default::<i32>();
+///
+/// assert_eq!(rect_map.get(Hex::new(0, 0)), Some(&0));
+/// ```
+///
+/// internally handle coordinate transform
+/// - `hex`
+/// - => `ij` offset coordinate
+/// - => `rc` 2D view of `Vec`
+/// - => `idx` `Vec` index
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "bevy_reflect", derive(bevy_reflect::Reflect))]
+#[cfg_attr(
+    feature = "bevy_ecs",
+    derive(bevy_ecs::resource::Resource, bevy_ecs::component::Component)
+)]
```

Comment:
```suggestion
```
For types other than Hex itself I think it's better to let users create their own wrappers for resources and components